docs/blog/gatekeeper-policies-as-oci-image.md

```diff
@@ -0,0 +1,215 @@
+      rego: |
+        package k8srequiredlabels
+        violation[{"msg": msg, "details": {"missing_labels": missing}}] {
+           provided := {label | input.review.object.metadata.labels[label]}
+           required := {label | label := input.parameters.labels[_].key}
+           missing := required - provided
+           count(missing) > 0
+           msg := sprintf("You must provide labels: %v for the %s: %s.", [missing, input.review.object.kind, input.review.object.metadata.name])
+        }
+        violation[{"msg": msg}] {
+           value := input.review.object.metadata.labels[key]
+           expected := input.parameters.labels[_]
+           expected.key == key
+           expected.allowedRegex != ""
+           not re_match(expected.allowedRegex, value)
+           msg := sprintf("Label %s: %s does not satisfy allowed regex: %s", [key, value, expected.allowedRegex])
+        }
+EOF
+```
+
+Define the associated `Constraint` for the `Namespaces` which must have the label `pod-security.kubernetes.io/enforce` as the key and either `baseline` or `restricted` as the value:
+```bash
+cat <<EOF> policies/ns-must-have-pss-label.yaml
+apiVersion: constraints.gatekeeper.sh/v1beta1
+kind: K8sRequiredLabels
+metadata:
+  name: ns-must-have-pss-label
+spec:
+  enforcementAction: deny
+  match:
+    kinds:
+      - apiGroups:
+        - ""
+        kinds:
+        - "Namespace"
+    excludedNamespaces:
+    - config-management-monitoring
+    - config-management-system
+    - gatekeeper-system
+    - kube-node-lease
+    - kube-public
+    - kube-system
+    - resource-group-system
+  parameters:
+    labels:
+    - key: pod-security.kubernetes.io/enforce
+      allowedRegex: (baseline|restricted)
+EOF
+```
+
+## Test this policy with local files
+
+Define a `Namespace` without the PSS label.
+```bash
+cat <<EOF > namespace-test.yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: test
+EOF
+```
+
+Let’s now locally test this `Namespace` against this policy with the `gator` CLI. This client makes it very convenient to test policies without any Kubernetes cluster!
+```bash
+gator test \
+   --filename namespace-test.yaml \
+   --filename policies/
+```
+
+Output similar to:
+```plaintext
+["ns-must-have-pss-label"] Message: "You must provide labels: {\"pod-security.kubernetes.io/enforce\"} for the Namespace: test."
+```
+
+## Push the Gatekeeper policy as OCI image to Artifact Registry
+
+Assuming we already have a [Google Artifact Registry repository](https://cloud.google.com/artifact-registry/docs/repositories/create-repos) `ARTIFACT_REGISTRY_REPO_NAME` in the region `REGION` and project `PROJECT_ID`.
+
+Login to Artifact Registry:
+```bash
+gcloud auth configure-docker ${REGION}-docker.pkg.dev
+```
+
+Push the Gatekeeper policy as OCI image in Google Artifact Registry repository with ORAS:
+```bash
+oras push \
+    ${REGION}-docker.pkg.dev/${PROJECT_ID}/${ARTIFACT_REGISTRY_REPO_NAME}/my-policies:1.0.0 \
+    policies/
```

Comment:
pushing without an artifactType or --config mediaType means the gatekeeper policy has no type. Or, the default of unknown

Suggest providing some type by adding:
`--config /dev/null:application/vnd.cncf.gatekeeper.v1`

Reply:
cc: @garethr for the actual mediaType as I know he'd done some work earlier.
Another suggestion would be to include the oras-go library into `gator` so you could do a `gator push|pull` directly. In this case, you could embed the appropriate `artifactType` directly into the gather cli.

Reply:
Thanks for your review @SteveLasker!
Yes indeed the `config.mediaType` is ``:
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.manifest.v1+json",
   "config": {
      "mediaType": "application/vnd.unknown.config.v1+json",
      "digest": "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a",
      "size": 2
   },
   "layers": [
      {
         "mediaType": "application/vnd.oci.image.layer.v1.tar+gzip",
         "digest": "sha256:325baa2355262c1cb5d8fb274759405f5c4b26e2110dfc22bbf9565740757938",
         "size": 1027,
         "annotations": {
            "io.deis.oras.content.digest": "sha256:6d2f0cfca84142481087d9566dd9e3a77a69d3e1a3754595c1615d09348487d8",
            "io.deis.oras.content.unpack": "true",
            "org.opencontainers.image.title": "policies"
         }
      }
   ],
   "annotations": {
      "org.opencontainers.image.created": "2023-02-08T17:11:52Z"
   }
}
```

Where this `mediaType` `application/vnd.cncf.gatekeeper.v1` is coming from and defined? Is it like a feature request you are suggesting for the Gatekeeper/`gator` project? Not seeing anything in that regard in there: https://open-policy-agent.github.io/gatekeeper/website/docs/gator/#bundling-policy-into-oci-artifacts.

Reply:
Regarding your other point about `oras-go` in `gator`, that's already the case: https://github.com/open-policy-agent/gatekeeper/pull/2398. That's what `gator test|expand --image` is leveraging, FYI.

Reply:
Also, @SteveLasker, is it blocking this PR to get your approval to be shipped in `main`? Maybe this should be a separate discussion in Slack or also, I'm more than happy to file an issue in Gatekeeper/`gator` repo to have this discussion there (depending on the answers of my questions above). Thanks!

Reply:
Thanks, @mathieu-benoit 
The mediaType usage is outlined here: [OCI Artifacts: Defining a Unique Artifact Type](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#defining-a-unique-artifact-type)

As for which `config.mediaType` to use, that should be defined by the OPA project, and that's why I was pinging @garethr, as I remember him doing so in a prior experiment. 

I wouldn't say it's blocking, rather the article does a great job and users tend to copy/paste examples, so I was hoping to add this clarity so everything in the registry didn't look "unknown". 
Here's another example [Visualizing Artifacts](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md#visualizing-artifacts), which is also supported by Docker Hub. See [Announcing Docker Hub OCI Artifacts Support](https://www.docker.com/blog/announcing-docker-hub-oci-artifacts-support/), with the visualization section title: **Viewing your Helm chart and using filters**

I see the additional gator support, which is awesome. It just looks like this `config.mediaType` was copied over as unknown, rather than identifying a well-known type.

Reply:
Gotcha, thanks for clarifying. 

I'll go ahead and resolve this discussion to unblock the review/shipment of this blog post. Hope that's ok?

As soon as the `config.mediaType` value to use is defined by the OPA/Gatekeeper project, it will be my pleasure to update that blog post accordingly.

Thanks!